ffi/webp.lua

```diff
@@ -0,0 +1,139 @@
+        -- When called from CreDocument:getImageFromPosition(), data is a "userdata" wrapping
+        -- a buffer malloc()'ed by cre.cpp, which is freed as soon as getImageFromPosition()
+        -- returns, as we usually don't need it after having rendered the image data to
+        -- a blitbuffer.
+        -- With animated multiframes images, we keep the decoder object: GifLib is fine with
+        -- data being gone, but libwebp is not.
+        -- We could have renderImageData() return a flag to state that the data should not
+        -- be free()d, but it would complexify the return values and their handling.
+        -- So, do simple: make a copy of the external buffer into an interned Lua string,
+        -- that we will keep alive.
+        libwebpdemux.WebPAnimDecoderDelete(decoder) -- give up with current decoder
+        data = ffi.string(data, size)
+        -- And call again this same function, now with data a real Lua string.
+        return Webp.fromData(data, size)
+    end
+
+    return Webp:new{
+        webp_decoder = decoder,
+        -- (We don't need to keep webp_data and dec_options alive, which just got
+        -- their values read and used and are not stored in the above objects.)
+        nb_frames = anim_info[0].frame_count,
+        width = anim_info[0].canvas_width,
+        height = anim_info[0].canvas_height,
+        -- we need to keep this interned Lua string alive as long as decoder is alive.
+        -- (To not be bothered with large binary data when using logger(webp), we don't
+        -- store the string itself, but a function referencing it as an upvalue, as
+        -- functions are just dumped as their address by logger.)
+        _input_data_holder = function() return data end,
+    }
+end
+
+function Webp:getFrameImage(number, no_copy)
+    if number < 1 then number = 1 end
+    if number > self.nb_frames then number = self.nb_frames end
+
+    if not self.webp_image then
+        self.webp_image = ffi.new("unsigned char*[1]")
+        -- WebPAnimDecoderGetNext() will update this pointer with the
+        -- address of one of its two internal bitmap data buffers.
+        -- It is so only valid until the next call.
+    end
+    if not self.cur_frame then
+        self.cur_frame = 0
+    end
+    if number < self.cur_frame then
+        libwebpdemux.WebPAnimDecoderReset(self.webp_decoder) -- Rewind to start
+        self.cur_frame = 0
+        -- Note: browsing frames forward is fast, but because of this, browsing
+        -- late frames backward can be noticably slower.
+    end
+    -- Render all frames from cur_frame to the one requested (this is needed as
+    -- webp frames can be partial and need to be blended over the previous frame)
+    local timestamp = ffi.new("int[1]") -- not used by us
+    while self.cur_frame < number do
+        local ret = libwebpdemux.WebPAnimDecoderGetNext(self.webp_decoder, self.webp_image, timestamp)
```

Comment:
I assume this means the blending of previous frames happens automagically, making the first buffer in `self.webp_image` valid?

Reply:
Yep, WebPAnimDecoder does it all for us. It uses other public objects in the library that deal with the lower level stuff and the real partial/alpha frames that are in the webp.
At least, that's what I understood from the documentation in the include/ files.

Reply:
Oh, that's cool (this sort of stuff is notoriously easy to get wrong when handling animated GIFs & PNGs, or so I heard ;p).